iceoryx_hoofs/include/iceoryx_hoofs/cxx/span.hpp

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_HOOFS_CXX_SPAN_HPP
+#define IOX_HOOFS_CXX_SPAN_HPP
+
+#include "iceoryx_hoofs/cxx/requires.hpp"
+#include "iceoryx_hoofs/cxx/span_iterator.hpp"
+#include "iceoryx_hoofs/cxx/type_traits.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <limits>
+#include <type_traits>
+#include <utility>
+
+using namespace std;
+
+namespace iox
+{
+namespace cxx
+{
+// constants
+constexpr uint64_t dynamic_extent = std::numeric_limits<uint64_t>::max();
+
+template <class From, class To>
+constexpr bool is_convertible_v = std::is_convertible<From, To>::value;
+
+// Implementation of C++17 std::size() and std::data()
```

Comment:
Q: where to put the implementation of C++17 std::size() and std::data()? Any recommendations? `iox::cxx:span` is not the correct place

Reply:
@mossmaurice if I understand [n4280](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4280.pdf) correctly they are part of `<iterator>`

Reply:
@patdhlk Ok. For use I would say they fit best in `type_traits.hpp`.

Reply:
@patdhlk In the STL, they are part of the individual container headers (vector.hpp etc.) and this is logical. We cannot have `type_traits` depend on those.

Reply:
@MatthiasKillat thank you for the input. I will leave them in the `span.hpp` and move the implementation to the `span.inl` file